### ISSUE TITLE
Wrong font spacing for some monospaced fonts

### ISSUE THREAD

**Issue by antanst** [tomszilagyi/zutty#67]
This is probably related to #37. For some fonts the spacing is completely off:

Fira Code:
![image](https://user-images.githubusercontent.com/966892/155487553-69a8df1e-157a-46f7-a083-1538c47bcf9f.png)

(Fira Mono looks fine though)

CartographCF:
![image](https://user-images.githubusercontent.com/966892/155487711-2b824b4e-c58c-4322-8a0d-2f762ded382d.png)

and others as well.

Having seen #37, maybe there is some kind of font metadata that is missing in those particular fonts, so Zutty may not be the one at fault here. They do look fine on other terminals I've tried though (Alacritty, GNOME Terminal)

Any ideas on a hack/workaround for those cases?

Thank you for the amazing software.

**Comment by tomszilagyi** [comment]
As also stated on #37, you need to use a fixed font variant (Fira Mono in this case). There is no point in using proportional type fonts in a terminal, and Zutty does not support them.

**Comment by antanst** [comment]
Perhaps I should have clarified that all the problematic fonts I mentioned are monospaced.

**Comment by tomszilagyi** [comment]
I installed and tested FiraCode with Zutty, and what I see in a newly added debug printout (in `Font::loadScaled ()` in src/font.cc) is that it reports the x and y size of glyphs (to be exact, `max_advance` and `height` fields that Zutty uses) as identical. Referring to https://freetype.org/freetype2/docs/reference/ft2-base_interface.html#ft_size_metrics the FT_Size_Metrics structure for FiraCode has:
```
  x_ppem = 16
  y_ppem = 16
  x_scale = 34415
  y_scale = 34415
  ascender = 960
  descender = -320
  height = 1280
  max_advance = 1280
```
This makes Zutty render the glyphs into squares, which looks incorrect, but is indeed correct based on the font metrics above. I do not know where else to get the more accurate (or in this case, the actual) glyph size information. You are welcome to investigate how other programs do it, then come back and tell me.

Failing to do that, as there are plenty of good monospace fonts that work fine in Zutty, I am not very concerned about this and have no interest in debugging this further (given that my resources are very limited). There will probably always be some font that is not correctly rendered for some reason or another. Zutty is a) probably not perfect (my apologies), and is b) less willing to go to great lengths to accommodate faults that lie elsewhere.

To summarize, you have a couple options:
1. use one of several fonts that work well with Zutty;
2. use a terminal that produces expected results with your chosen fonts;
3. debug the issue and let me know what you find.

**Comment by antanst** [comment]
Thank you for the detailed explanation and for your time.

**Comment by clort81** [comment]
> I installed and tested FiraCode with Zutty, and what I see in a newly added debug printout (in `Font::loadScaled ()` in src/font.cc) is that it reports the x and y size of glyphs (to be exact, `max_advance` and `height` fields that Zutty uses) as identical. Referring to https://freetype.org/freetype2/docs/reference/ft2-base_interface.html#ft_size_metrics the FT_Size_Metrics structure for FiraCode has:
> 
> ```
>   x_ppem = 16
>   y_ppem = 16
>   x_scale = 34415
>   y_scale = 34415
>   ascender = 960
>   descender = -320
>   height = 1280
>   max_advance = 1280
> ```
> 
> This makes Zutty render the glyphs into squares, which looks incorrect, but is indeed correct based on the font metrics above. I do not know where else to get the more accurate (or in this case, the actual) glyph size information. You are welcome to 

Some glyphs in some monospace fonts are 'double wide' so the max advance will reflect 2x the desired width.  
Libvte scans through all glyphs to get an actual width.
The .bdf format has per-glyph width DWIDTH.
